### ISSUE TITLE
there is a small isuue on windows which can casue crash

### ISSUE THREAD

**Issue by jellyr** [lenmus/lomse#99]
while (i < m_totalGlyphs && width + nextWidth < maxSpace)
    {
        if (m_glyphs[i] == L' ')
        {
            if (!fSpaces)
            {
                breakPoint = i;
                fSpaces = true;
            }
        }
        else
            fSpaces = false;

        width += nextWidth;
        nextWidth = m_glyphWidths[++i];  // in the final loop,  when i is m_totalGlyphs-1, ++i is m_totalGlyphs, which casuse m_glyphWidths overflow。   it should be i++
    }

**Comment by cecilios** [comment]
Thanks for reporting.

Just for future reference, this code is in lomse_text_splitter.cpp, line 119.

Now, returning to the code, I think it is correct to use ++i as we are trying to get the width of next element. But it is true that when i is m_totalGlyphs-1, accessing m_glyphWidths[++i] will access out of vector size. 

In Unit Tests I didn't noticed any problem, so investigating this it seems that std::vector operator [] won't throw any out of range error and it is an undefined behaviour to access element greater than the vector size using [] operator. Probably it is safe in Linux but not in Windows, but anyway it should be fixed.

Also code in ChineseTextSplitter::get_next_text_engrouter()  [same file, line 198] has to be fixed.

I would appreciate if you could submit a pull request with a patch for this issues. Thank you.

**Comment by jellyr** [comment]
I'd suggest to change
` while (i < m_totalGlyphs && width + nextWidth < maxSpace)`
to 
 `while (i < m_totalGlyphs-1 && width + nextWidth < maxSpace)`

I haven't got confidence to change your code , bcz I  haven't got big picture about your code . 

I just want to change your render engine to "2D Vector Graphics Library" as follows .
https://www.shadertoy.com/view/lslXW8
bcz , there is so much cool effect in this website that I can apply to the app :)

**Comment by cecilios** [comment]
> I'd suggest to change
> while (i < m_totalGlyphs && width + nextWidth < maxSpace)
> to
> while (i < m_totalGlyphs-1 && width + nextWidth < maxSpace)

No, that is not correct. The fix is to avoid accessing the vector when index is grater than vector size. I would suggest to change lines 133 and 198 from:

```
        nextWidth = m_glyphWidths[++i];
```
to:

```

	if (++i == m_totalGlyphs)
	    break;
        nextWidth = m_glyphWidths[i];
```

It will be great if you could prepare this PR. I'm quite busy now in the middle of a big PR for adding more features and I wouldn't like to postpone this simple fix, for two or three weeks, until I finish my current PR. And will be a way of giving credit to you in the git changelog. Thank you!

> I just want to change your render engine to "2D Vector Graphics Library" as follows .
> https://www.shadertoy.com/view/lslXW8
> bcz , there is so much cool effect in this website that I can apply to the app :)

I do not understand very well what are you trying to do. The code used by Lomse for rendering, the AGG library (https://en.wikipedia.org/wiki/Anti-Grain_Geometry) is very powerful and is the base for many other 2D Vector Graphics Libraries. See the samples to get and idea of the effects you can easily get (http://antigrain.com/screenshots/index.html).

As you point me to shadertoy.com, perhaps what you are trying to do is to make changes to allow direct rendering in web pages. This will be a great enhancement for the Lomse library but will not require to replace the current AGG rendering code, but just to create a new class derived from Drawer. For instance, a Drawer for generating SVG code, or perhaps a Drawer based on WebGL or on ShaderToy.

**Comment by cecilios** [comment]
@jellyr  To save time I have added the fix in PR #100
